prac_07/project_management.py

```diff
@@ -0,0 +1,178 @@
+"""
+CP1404 practical 7 - Project management program (do from scratch)
+
+estimate: 90 minutes (no idea!)
+actual time: honestly don't know (was somewhat on track at around 55 minutes, then got fixated on some details and ended
+up taking way more time.).
+
+notes/questions:
+- Wasn't sure about using modules inside classes, so currently storing start_date as a string (but probably shouldn't).
+- so many functions for getting numbers... seems unnecessary (also inconsistent approach to range check for the
+percentage in update project section).
+(- Repetitive exception handling? (but also missing FileNotFound exception))
+"""
+from datetime import datetime
+from prac_07.project import Project
+
+FILENAME = "projects.txt"
+
+MENU = """- (L)oad projects  
+- (S)ave projects  
+- (D)isplay projects  
+- (F)ilter projects by date
+- (A)dd new project  
+- (U)pdate project
+- (Q)uit"""
+
+
+def main():
+    """Program to track and update project list files."""
+    projects, headers = load_projects(FILENAME)
+    print(MENU)
+    menu_choice = input(">>>").upper()
+    while menu_choice != "Q":
+        if menu_choice == "L":
+            filename = input("Filename: ")
+            projects = load_projects(filename)
+        elif menu_choice == "S":
+            filename = input("Filename: ")
+            save_projects(projects, filename, headers)
+        elif menu_choice == "D":
+            display_projects(projects)
+        elif menu_choice == "F":
+            filter_projects(projects)
+        elif menu_choice == "A":
+            add_project(projects)
+        elif menu_choice == "U":
+            project_choice = get_project_choice(projects)
+            update_project(project_choice)
+            projects.sort()
+        else:
+            print("Invalid input!")
+        print(MENU)
+        menu_choice = input(">>>").upper()
+    save_projects(projects, "temporary_file.txt", headers)
+    print("Thank you for using custom-built project management software.")
+
+
+def load_projects(filename):
+    """Load list of Projects from file."""
+    projects = []
+    with open(filename, 'r', encoding="UTF-8") as in_file:
+        headers = in_file.readline().strip()
+        for line in in_file:
+            parts = line.strip().split("\t")
+            project = Project(parts[0], parts[1], int(parts[2]), float(parts[3]), int(parts[4]))
+            projects.append(project)
+    projects.sort()
```

Comment:
Do the projects need to be sorted by priority here or only when displaying completed and incomplete projects? They will be resorted again when filtering by date so not sure. This could be the better approach though.

Reply:
Good question, I'm not sure. I guess it depends on whether you just want the list to be sorted for the display option, or also in general. Why are they are resorted when filtering by date?